### ISSUE TITLE
marco window foreground/background behaviour broken when switching workspaces

### ISSUE THREAD

**Issue by pasikarkkainen** [mate-desktop/marco#560]
With the default Mate/Marco settings "Enable software compositing window manager" -> "enabled" Marco is very problematic/broken when switching workspaces. When you have many windows open on a workspace, and you switch between different workspaces (either with a mouse from the workspace switcher applet, or with a keyboard shortcut), it seems some random window gets the focus but the focused window is not on the foreground, instead it's somewhere on the background with all the other (or at least most of the other) windows in front of it.. 

1) you can't see the focused window properly, because it's in the background, all the other / most of the other windows are being in front of it / "hiding" it.
2) you can't click/interact with what you see on the desktop/workspace, because all the input goes to that weird/broken focused window in the background, not to that window which you see and try to interact with..

Restoring the focused window to foreground is very tricky, because all the focus goes to this broken focused window in the background, so you can't easily minimize the other windows if/when they're overlapping, because like said the focus goes to that weird focused window in the background instead of the window you try to minimize.. easiest way to get the focused window to foreground is to click the "hide all windows"-button, and then find the wanted window to show it.

Disabling the "Enable software compositing window manager" setting fixes the problematic behaviour.

This feels/seems like exactly the same issue as a previous marco issue ( https://github.com/mate-desktop/marco/issues/251 ). copy/paste from the previous issue:

"windows is not brought to the front, it stays in the back. So, it looks like I'm clicking on the correct window, but I'm clicking in one in the background."

#### Expected behaviour
Focused window is on the foreground after switching workspaces.

#### Actual behaviour
Focused window goes to background, with many (if not all?) windows in front of it.. so you can't see the focused window, except all the input goes to that weird/broken window in the background.

#### Steps to reproduce the behaviour
Use mate/marco with the default settings (software compositing window manager -> enabled), have many windows open on each workspace, and switch between workspaces a few times.

#### MATE general version
mate 1.22

#### Package version
marco-1.22.3-1.fc30.x86_64

#### Linux Distribution
Fedora 30.

I'm seeing this same broken behaviour on multiple systems: on a single-display laptop, and also on a multi-display workstation.

**Comment by raveit65** [comment]
Does this happens with marco-1.22.1-1.fc30 ?
https://koji.fedoraproject.org/koji/buildinfo?buildID=1256952
Which GPU and driver are you using?

Honestly, i can't reproduce it with nvidia GPU and proprietary nvidia driver.
To avoid issues i don't use old radeon/amd/intel graphic GPUs.

**Comment by vkareh** [comment]
I've actually managed to reproduce this occasionally on _one_ of my laptops (not the other), but I don't have a clear pattern on how/why it occurs. I use the nvidia proprietary driver on that machine, with marco built from master on F30. Now that someone else has reported it, I might start looking at the root cause.

**Comment by raveit65** [comment]
I use those settings and i can't reproduce it. Maybe this helps?
```
[rave@mother ~]$ gsettings get org.mate.Marco.general auto-raise
true
[rave@mother ~]$ gsettings get org.mate.Marco.general auto-raise-delay
800
[rave@mother ~]$ gsettings get org.mate.Marco.general focus-mode
'sloppy'
```
In result the window gets always focus when i am hovering with the mouse over the window.

**Comment by pasikarkkainen** [comment]
@raveit65 I'm using Intel integrated gfx on all the problematic systems. And yeah, I don't use/want 'auto-raise'.. the problem is 100% reproducible with the *default* mate/marco settings. But obviously you need to have many windows open on each workspace to reproduce. More windows you have the easier it is to reproduce.

**Comment by lukefromdc** [comment]
Probably I never have enough windows open while testing marco to see this, as I use compiz for most other work. I agree that focus follows mouse (auto-raise) is unsuitable for many users, myself included.  Also I do not use proprietary video drivers or cards "requiring" them for reasons of trust so can't test under that condition. Odd though that a CPU renderer would have a focus issue affected by a GPU driver bug or difference though.

OK, someone familiar with the part of marco's code in question could use as a starting point the fact that focus follows mouse works, though that might just be automatically SHIFTING focus to a new window. What happens if when this bug is showing, you pick an arbitrary window and click on it?  The expected behavior would be that window focussed, same as mousing over a window under auto-raise.

**Comment by raveit65** [comment]
@pasikarkkainen 

> And yeah, I don't use/want 'auto-raise'..

Ok, but how is the result if you do something which you don't want ? :)
I mean we want to debug this issue. So saying i don't want to test a setting makes it really complicate.
In result i don't know why i can't reproduce your issue.....and i can go in holidays...... \o/

**Comment by raveit65** [comment]
> I'm using Intel integrated gfx on all the problematic systems.

???? no idea what you`re want to say or using....
Intel driver in fedora called `xorg-x11-drv-intel`

**Comment by lukefromdc** [comment]
When when you don't normally use a setting, you can TEST it unless it would create a problem for you preventing you from reversing that change e.g changing to a language you cannot understand well enough to find your way back.

**Comment by pasikarkkainen** [comment]
Of course I can TEST whatever settings :) I just meant that the Mate/Marco defaults are broken, resulting in the behaviour described in this issue.

**Comment by raveit65** [comment]
@pasikarkkainen 
And you didn't answer another important question.
Again, does this happens with marco-1.22.1-1.fc30 ?
https://koji.fedoraproject.org/koji/buildinfo?buildID=1256952

**Comment by pasikarkkainen** [comment]
@raveit65 yep I'll test with the older version and let you know how it goes. (I've been away for a few days now so I haven't been able to test yet).

**Comment by pasikarkkainen** [comment]
@raveit65 I've now downgraded to marco-1.22.1-1.fc30. I'll report the experiences here.

Something I realized only now is the fact that the problematic behaviour doesn't happen right after logging in, but instead starts after an hour or two of use.. maybe there's some kind of event that triggers it. Anyway, I'll test the older version of marco now and let you know how it goes.

**Comment by lukefromdc** [comment]
Keep an eye on marco's RAM use as you run a session, see if reaching some level of increased RAM usage correlates to this. That would indicate a memory leak combined with some kind of corruption, or not freeing all data related to closed windows and eventually reaching a threshold of open windows plus unfreed remnants from previously open windows holding open gtk objects that normally would be destroyed when the windows are closed. That's assuming it isn't just taking an hour or two to reach the trigger number of open windows.

**Comment by pasikarkkainen** [comment]
@lukefromdc : ok, will do.

I've now used the older marco-1.22.1-1.fc30 version for some days on my secondary laptop, so it's not the daily driver, but still I've now used it long enough to say that marco-1.22.1-1.fc30 behaves better than the newer marco-1.22.3-1.fc30.x86_64 version. I haven't seen the buggy behaviour yet with the older 1.22.1 marco version.

With the newer marco-1.22.3-1.fc30.x86_64 I would have already gotten the buggy behaviour many times at this point.. so yes, the older 1.22.1 version of marco seems to work better, I'd say.

I'll keep you updated if I happen to get the buggy behaviour with the older version..

**Comment by raveit65** [comment]
So, marco-1.22.1-1.fc30 doesn't have the issue.
@pasikarkkainen 
Did you notice the issue with marco-1.22.2-1.fc30 ?
This would help us to determine which commit(s) could causes the issue.
Thanks.

**Comment by pasikarkkainen** [comment]
@raveit65 I'll switch to using marco-1.22.2-1.fc30 and see how that goes.

**Comment by Chasx** [comment]
I have the same problem on two computers of mine running Linux Mint Mate 17.**3** (based on ubuntu 18.04). The version number of Marco is '1.22.2-1+tina'.

I have seen the problem before (a couple of years back) but it was gone for a while.

I think it MIGHT also not be present in Linux Mint Mate 17.**2** on my laptop (also Ubuntu 18.04 based but which is running Marco '1.20.1-2ubuntu1').

**Comment by raveit65** [comment]
Yesterday it was the first time that i did run in this issue when i use the default settings of marco.
So it it pretty clear for me that it only happens when i **don't** use my normal settings.
```
[rave@mother ~]$ gsettings get org.mate.Marco.general auto-raise
true
[rave@mother ~]$ gsettings get org.mate.Marco.general auto-raise-delay
800
[rave@mother ~]$ gsettings get org.mate.Marco.general focus-mode
'sloppy'
```
Now i am testing only with using `focus-mode 'sloppy'`.

**Comment by Chasx** [comment]
Not sure if it helps but I have the problem with the following settings in Linux Mint Mate (which I think are default?):
```
auto-raise: false
auto-raise-delay: 500
focus-mode: 'click'
```

**Comment by raveit65** [comment]
Yes, that are the default setting.
With that settings i never had this issue since i used invisible-border feature from june this year.
I suspect that it only happens when using the default settings.

**Comment by raveit65** [comment]
@pasikarkkainen 
Can you please test this build (f30 and f31) with a possible fix.
https://koji.fedoraproject.org/koji/taskinfo?taskID=39414126
https://koji.fedoraproject.org/koji/taskinfo?taskID=39414128

@Chasx 
Try to use this commit https://patch-diff.githubusercontent.com/raw/mate-desktop/marco/pull/570.patch

**Comment by vkareh** [comment]
This should now be fixed on `master` by #570 and cherry-picked to `1.22`: baef3d7f25bbd4665ecced218be3b4e97c4d8a02
